core/shared/src/main/scala/scalaz/zio/Fiber.scala

```diff
@@ -63,6 +74,26 @@ trait Fiber[+E, +A] { self =>
 
       def interrupt0(ts: List[Throwable]): IO[Nothing, Unit] =
         self.interrupt0(ts) *> that.interrupt0(ts)
+
+      def finished(fc: ExitResult[E1, C] => IO[Nothing, Unit]): IO[Nothing, Unit] =
+        self.finished { ra: ExitResult[E, A] =>
+          that.finished { rb: ExitResult[E1, B] =>
+            (ra, rb) match {
+              case (ExitResult.Completed(a), ExitResult.Completed(b)) => fc(ExitResult.Completed(f(a, b)))
+              case _ =>
+                ra match {
+                  case ExitResult.Failed(e, ts)  => fc(ExitResult.Failed(e, ts))
+                  case ExitResult.Terminated(ts) => fc(ExitResult.Terminated(ts))
+                  case _                         => IO.unit
+                }
+                rb match {
```

Comment:
You need to stitch these two actions together, otherwise you're throwing away the result of `ra match`.

Add `*>` right after the close of `ra match { ... }`, i.e.:

```scala
ra match {
...
} *> 
rb match {
...
}
```

Reply:
🤦‍♂️